crates/red_knot/src/symbols.rs

```diff
@@ -557,10 +557,13 @@ impl<'a> Iterator for ReachableDefinitionsIterator<'a> {
             match &self.table.flow_graph.flow_nodes_by_id[self.flow_node_id] {
                 FlowNode::Start => return None,
                 FlowNode::Definition(def_node) => {
-                    self.flow_node_id = def_node.predecessor;
                     if def_node.symbol_id == self.symbol_id {
+                        // we found a definition; previous definitions along this path are not
+                        // reachable
+                        self.flow_node_id = FlowGraph::start();
                         return Some(def_node.definition.clone());
```

Comment:
Nit: Maybe add an empty line in between to make it clear that the comment only applies to the assignment
```suggestion
                        self.flow_node_id = FlowGraph::start();
                        
                        return Some(def_node.definition.clone());
```

Reply:
Point taken, but this code changes in next PR anyway; not going to change it here which will just cause conflicts.